oauth/scope.js

```diff
@@ -0,0 +1,212 @@
+    }
+  }
+
+  toString() {
+    return this.url.href;
+  }
+
+  implies(value) {
+    if (! value.url) {
+      return false;
+    }
+    if (value.url.origin !== this.url.origin) {
+      return false;
+    }
+    if (value.url.pathname !== this.url.pathname) {
+      if (! value.url.pathname.startsWith(this.url.pathname + '/')) {
+        return false;
+      }
+    }
+    if (this.url.hash) {
+      if (value.url.hash !== this.url.hash) {
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+
+class Scope {
+  constructor(values) {
+    this.values = values;
+  }
+
+  toString() {
+    return this.values.map(value => {
+      return value.toString();
+    }).join(' ');
+  }
+
+  /**
+   * Check whether one scope implies another.
+   *
+   * A scope A implies another scope B if every scope value
+   * in B is implied by some scope value in A.  In other words,
+   * A represents a strictly more general set of capabilities
+   * than B.
+   *
+   */
+  implies(scope) {
+    return scope.values.every(target => {
+      return this.values.some(value => {
+        return value.implies(target)
+      });
+    });
+  }
+
+  /**
+   * Filter a scope by the capabilities of another.
+   *
+   * Filtering a scope A by another scope B will return the
+   * largest subset of scope values from A that are implied
+   * by B.
+   *
+   * It's useful to think of `A.filter(B)` as checking a set
+   * a set of requested capabilities in `A` against a set of
+   * allowed capabilities in `B`.  The result will be the subset
+   * of requested capabilities that are actually allowed.
+   *
+   */
+  filter(scope) {
+    return new Scope(this.values.filter(ask => {
+      return scope.values.some(allow => {
+        return allow.implies(ask)
+      });
+    }));
+  }
+
+}
+
+
+module.exports = {
+
+  /**
+   * Parse a space-delimited string into a Scope object.
+   *
+   * This function implements the semantics defined in RFC6749,
+   * where the "scope" input string represents a space-delimited
+   * list of case-sensitive strings identifying individual scopes.
+   *
+   */
+  fromString (scopeString) {
+    return new Scope(scopeString.split(/ +/).filter(valueString => {
+      return !! valueString
+    }).map(valueString => {
+      if (valueString.startsWith('https:')) {
+        return new URLScopeValue(valueString);
+      } else {
+        return new ShortScopeValue(valueString);
+      }
+    }));
+  },
+
+  /**
+   * Parse a url-encoded string into a Scope object.
+   *
+   * This function parses a ScopeSet from a "scope" input string
+   * that has been url-encoded, as you might receive in the query
+   * parameter of an OAuth authorizaton request.  The list is thus
+   * delimited by `+` rather than ` `, and any percent-encoded
+   * characters in the scopes will be expanded.
+   *
+   */
+  fromURLEncodedString (encodedScopeString) {
+    return new Scope(encodedScopeString.split(/\+/).filter(encodedValueString => {
```

Comment:
above we have `split(/ +/)` and here we have `split(/\+/)`, which to use?

Reply:
They're different on purpose, the first is "split on one or more spaces" and the second is "split on a plus symbol"

Reply:
right, would it make sense to document that or too obvious? In the past there were "deprecated escape chars" in regexes, i think that is why this alerted me